sklearn/cross_decomposition/pls_.py

```diff
@@ -375,14 +375,14 @@ def transform(self, X, Y=None, copy=True):
         x_scores if Y is not given, (x_scores, y_scores) otherwise.
         """
         check_is_fitted(self, 'x_mean_')
-        X = check_array(X, copy=copy)
+        X = check_array(X, copy=copy, dtype=FLOAT_DTYPES)
         # Normalize
         X -= self.x_mean_
         X /= self.x_std_
         # Apply rotation
         x_scores = np.dot(X, self.x_rotations_)
         if Y is not None:
-            Y = check_array(Y, ensure_2d=False, copy=copy)
+            Y = check_array(Y, ensure_2d=False, copy=copy, dtype=FLOAT_DTYPES)
             if Y.ndim == 1:
                 Y = Y.reshape(-1, 1)
             Y -= self.y_mean_
```

Comment:
@GaelVaroquaux @ogrisel @lesteve I am also wondering if instead of explicitly typecasting `Y` to be `float` (assuming `y_mean_` is `float`) should we -

1. typecast `y_mean_` to `Y.dtype` or
2. typecast `Y` to `y_mean_.dtype`?